src/ad/primitives.jl

```diff
@@ -165,35 +165,108 @@ end
 function backward!(i::BroadcastInstruction{<:Tuple{typeof(*),Any,Any}})
     _, x, y = i.input
     z = i.output
-    x.deriv .+= value(y) .* deriv(z)
-    y.deriv .+= value(x) .* deriv(z)
+    @propagate!(x, value(y) .* deriv(z))
+    @propagate!(y, value(x) .* deriv(z))
     return nothing
 end
 
 function backward!(i::BroadcastInstruction{<:Tuple{typeof(+),Any,Any}})
     _, x, y = i.input
     z = i.output
-    x.deriv .+= deriv(z)
-    y.deriv .+= deriv(z)
+    @propagate!(x, deriv(z))
+    @propagate!(y, deriv(z))
     return nothing
 end
 
 #=== mixed-mode broadcast optimization ===#
 
 @inline inbounds_partials(d, i) = @inbounds ForwardDiff.partials(d, i)
 
-@inline function backprop_partial(input_deriv, output_dual, ::Val{i}, output_deriv) where i
-    return input_deriv + (inbounds_partials(output_dual, i) * output_deriv)
+@inline function backprop_partial(output_dual, i, output_deriv)
+    return inbounds_partials(output_dual, i) * output_deriv
+end
+
+@inline function multivariable_backprop_partial(input_deriv, output_dual, i, output_deriv)
+    return input_deriv + backprop_partial(output_dual, i, output_deriv)
 end
 
 function backward!(i::BroadcastInstruction)
     f, args = first(i.input), i.input[2:end]
-    output, output_duals = i.output
-    output_deriv = deriv(output)
-    for (i, arg) in enumerate(args)
-        isa(arg, Variable) || continue
-        arg_deriv = deriv(arg)
-        broadcast!(backprop_partial, arg_deriv, arg_deriv, output_duals, Val(i), output_deriv)
-    end
+    output_variable, output_duals = i.output
+    dual_broadcast_backward!(args, output_duals, deriv(output_variable))
     return nothing
 end
+
+## CPU impl
+
+@generated function dual_broadcast_backward!(inputs::NTuple{N,Any}, output_duals,
+                                             output_derivs) where {N}
+    body = Expr(:block)
+    for i in 1:N
+        if inputs.parameters[i] <: Variable
+            input = :(inputs[$i])
+            push!(body.args, quote
+                if $input.downstreams > 1
+                    broadcast!(multivariable_backprop_partial, deriv($input), deriv($input),
+                               output_duals, $i, output_derivs)
+                else
+                    broadcast!(backprop_partial, deriv($input), output_duals, $i,
```

Comment:
Ok so instead of specialising on `Val` we are now relying on inlining and const prop?

Reply:
yeah.